src/Tasks/Microsoft.NET.Build.Tasks/ResolveTargetingPackAssets.cs

```diff
@@ -133,12 +225,23 @@ protected override void ExecuteCore()
                 }
             }
 
-            //  Remove RuntimeFramework items for shared frameworks which weren't referenced
+            //  Calculate which RuntimeFramework items should actually be used based on framework references
             HashSet<string> frameworkReferenceNames = new HashSet<string>(FrameworkReferences.Select(fr => fr.ItemSpec), StringComparer.OrdinalIgnoreCase);
-            RuntimeFrameworksToRemove = RuntimeFrameworks.Where(rf => !frameworkReferenceNames.Contains(rf.GetMetadata(MetadataKeys.FrameworkName)))
-                                        .ToArray();
+            UsedRuntimeFrameworks = RuntimeFrameworks.Where(rf => frameworkReferenceNames.Contains(rf.GetMetadata(MetadataKeys.FrameworkName)))
+                                    .ToArray();
 
-            ReferencesToAdd = referencesToAdd.ToArray();
+            //  Filter out duplicate references (which can happen when referencing two different profiles that overlap)
+            HashSet<string> seenReferences = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            List<TaskItem> deduplicatedReferences = new List<TaskItem>(referencesToAdd.Count);
+            foreach (var reference in referencesToAdd)
+            {
+                if (seenReferences.Add(reference.ItemSpec))
+                {
+                    deduplicatedReferences.Add(reference);
+                }
```

Comment:
I think Distinct() preserves order in practice, but docs say unordered so maybe not wise to use it here. I think this should be factored into a helper if we can't use it.